pytorch_lightning/trainer/trainer.py

```diff
@@ -1082,6 +1083,15 @@ def call_setup_hook(self, model: LightningModule) -> None:
         self.setup(model, stage=state)
         model.setup(stage=state)
 
+    def call_configure_sharded_model(self, model: LightningModule) -> None:
+        # Call configure sharded model hook if accelerator requests. In some cases
+        # we will not call the hook; the hook has initialized the sharded model for example.
+        if self.accelerator.call_configure_sharded_model_hook:
+            with self.accelerator.model_sharded_context():
+                model.configure_sharded_model()
+                self.configure_sharded_model(model)
+            self.accelerator.call_configure_sharded_model_hook = False
+
```

Comment:
@SeanNaren @kaushikb11 why's this needed in the trainer? could this be part of the accelerator setup? in L440 instead?

Reply:
which part exactly? the trainer has control over the hook, we do not want the hook to be called within the accelerator i think

Reply:
I see, on first read I thought that we could do this:
```
self.call_setup_hook(model)
self.accelerator.setup(self, model)  -> inside here apply the model sharded context and call model.configure_sharded_model()

# now that accelerator has finished doing the model sharding setup call the callbacks
self.configure_sharded_model(model)

but this splits the callback hook from the setup which can be less convenient

Reply:
Agree as long as we're ok with calling the model hook within the accelerator (cc @justusschock @awaelchli)